nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/main/java/org/apache/nifi/processors/GeoEnrichIPRecord.java

```diff
@@ -0,0 +1,323 @@
+    protected boolean splitOutput;
+
+    @Override
+    @OnScheduled
+    public void onScheduled(ProcessContext context) throws IOException {
+        super.onScheduled(context);
+
+        readerFactory = context.getProperty(READER).asControllerService(RecordReaderFactory.class);
+        writerFactory = context.getProperty(WRITER).asControllerService(RecordSetWriterFactory.class);
+        splitOutput = context.getProperty(SPLIT_FOUND_NOT_FOUND).asBoolean();
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (input == null) {
+            return;
+        }
+
+        FlowFile output = session.create(input);
+        FlowFile notFound = splitOutput ? session.create(input) : null;
+        final DatabaseReader dbReader = databaseReaderRef.get();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             OutputStream osNotFound = splitOutput ? session.write(notFound) : null) {
+            RecordPathCache cache = new RecordPathCache(GEO_PROPERTIES.size() + 1);
+            Map<PropertyDescriptor, RecordPath> paths = new HashMap<>();
+            for (PropertyDescriptor descriptor : GEO_PROPERTIES) {
+                if (!context.getProperty(descriptor).isSet()) {
+                    continue;
+                }
+                String rawPath = context.getProperty(descriptor).evaluateAttributeExpressions(input).getValue();
+                RecordPath compiled = cache.getCompiled(rawPath);
+                paths.put(descriptor, compiled);
+            }
+
+            String rawIpPath = context.getProperty(IP_RECORD_PATH).evaluateAttributeExpressions(input).getValue();
+            RecordPath ipPath = cache.getCompiled(rawIpPath);
+
+            RecordReader reader = readerFactory.createRecordReader(input, is, getLogger());
+            RecordSchema schema = writerFactory.getSchema(input.getAttributes(), null);
+            RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, os);
+            RecordSetWriter notFoundWriter = splitOutput ? writerFactory.createWriter(getLogger(), schema, osNotFound) : null;
+            Record record;
+            Relationship targetRelationship = REL_NOT_FOUND;
+            writer.beginRecordSet();
+
+            while ((record = reader.nextRecord()) != null) {
+                CityResponse response = geocode(ipPath, record, dbReader);
+                boolean wasEnriched = enrichRecord(response, record, paths);
+                if (wasEnriched) {
+                    targetRelationship = REL_FOUND;
+                }
+                if (!splitOutput || (splitOutput && wasEnriched)) {
+                    writer.write(record);
+                } else {
+                    notFoundWriter.write(record);
+                }
+            }
+            writer.finishRecordSet();
```

Comment:
You've flushed and closed `writer` here but not `notFoundWriter`, I believe this can lead to missing/incorrect output. I found it when noticing the comment below about the `notFound` flow file being sent out with 0 records, but I used a JSONRecordSetWriter so I should have seen at least two bytes (an empty array) in the output, but it was zero bytes.